### ISSUE TITLE
RFC: Deepen training orchestration — unified buffer, trainer, config bridge

### ISSUE THREAD

**Issue by LSailer** [LSailer/Master-Thesis-3D-VLA#68]
## Problem

The training pipeline has three scripts that are ~70% copy-pasted:

| Script | Lines | Buffer | Configs created | `_convert_batch` | `_save_checkpoint` |
|---|---|---|---|---|---|
| `run_jax_crafter.py` | 139 | `ReplayBuffer` | 1 (`R2DreamerConfig`) | copy #1 | none |
| `run_jax_habitat.py` | 302 | `ReplayBuffer` | 2 (`R2DreamerConfig` + `DreamerConfig`) | copy #2 | copy #1 |
| `run_jax_habitat_vggt.py` | 317 | `VGGTReplayBuffer` | 2 + extractor | copy #3 | copy #2 |

**Shallow modules:**
- `ReplayBuffer` and `VGGTReplayBuffer` are 95% identical — differ only in storage dtype (`uint8` vs `float32`) and whether to normalize (`/255`) on sample
- `_convert_batch()` (replay buffer output → agent input) is copy-pasted 3 times
- `_save_checkpoint()` is copy-pasted 2 times — and both copies **miss `agent.ema_state`**, so resuming training resets return normalization

**Coupled configs:**
- `HabitatObjectNavEnv` requires `DreamerConfig`; `R2DreamerAgent` requires `R2DreamerConfig`. Scripts construct both, even though fields like `obs_shape`, `num_actions`, `buffer_capacity` overlap.

**Training loop duplication:**
- Prefill → train (with train-ratio credit) → log (CSV + optional WandB) → checkpoint — the same structure in every script, with env-specific episode metrics interleaved

## Proposed Interface

Two changes: a unified `ReplayBuffer` and a new `modules/r2dreamer/trainer.py`.

### 1. Unified ReplayBuffer (`modules/dreamerv3/replay_buffer.py`)

```python
@dataclass(frozen=True)
class BufferConfig:
    capacity: int
    obs_shape: tuple[int, ...]       # (3,64,64) or (4116,)
    obs_dtype: str = "uint8"         # "uint8" or "float32"
    normalize_obs: bool = True       # /255 on sample (only meaningful for uint8)

class ReplayBuffer:
    def __init__(self, config: BufferConfig) -> None: ...
    def add(self, obs, action, reward, done, terminal=False) -> None: ...
    def sample(self, batch_size, seq_len) -> dict: ...
```

Replaces both `ReplayBuffer` and `VGGTReplayBuffer`. `ValReplayDataset` stays as-is.

### 2. Trainer module (`modules/r2dreamer/trainer.py`)

```python
@dataclass
class TrainerConfig:
    output_dir: str
    total_steps: int = 10_000_000
    prefill_steps: int = 5000
    log_every: int = 250
    checkpoint_every: int = 50_000
    seed: int = 0
    wandb_project: str | None = "3d-vla-objectnav"
    wandb_name: str | None = None
    wandb_tags: list[str] = field(default_factory=lambda: ["r2dreamer"])
    val_data: str | None = None
    val_loss_every: int = 10_000

@dataclass
class ObsAdapter:
    """Bridges env observations to agent/buffer. Called once per step."""
    buffer_dtype: np.dtype = np.uint8
    buffer_shape: tuple[int, ...] = (3, 64, 64)
    normalize_on_sample: bool = True
    on_episode_reset: Callable[[], None] | None = None

    def transform(self, obs_dict: dict) -> tuple[np.ndarray, dict]:
        """Returns (buffer_obs, agent_obs_dict). Default: image for buffer, pass-through for agent."""
        return obs_dict["image"], obs_dict

def convert_batch(batch: dict, num_actions: int) -> dict:
    """Replay buffer output → agent input. One-hot actions, dones→is_last, terminals→is_terminal."""

def save_checkpoint(agent, step: int, output_dir: str) -> str:
    """Saves params, opt_state, slow_critic_params, AND ema_state. Returns path."""

def load_checkpoint(agent, path: str) -> int:
    """Restores agent state. Returns step number."""

def habitat_defaults(env) -> dict:
    """Returns dict(obs_adapter=..., episode_metrics_fn=...) for the Habitat+CNN case."""

class Trainer:
    def __init__(self, agent, env, agent_config, trainer_config: TrainerConfig,
                 obs_adapter: ObsAdapter = ObsAdapter(),
                 episode_metrics_fn: Callable | None = None) -> None: ...
    def run(self) -> None:
        """Prefill → train loop → final checkpoint. Blocks until done."""
```

### Usage — Habitat (the 80% case, ~50 lines)

```python
config = R2DreamerConfig(obs_shape=(3,64,64), num_actions=4, ...)
env = HabitatObjectNavEnv(hab_config, ...)
agent = R2DreamerAgent(config, init_key)

trainer = Trainer(
    agent=agent, env=env, agent_config=config,
    trainer_config=TrainerConfig(output_dir=args.output_dir, wandb_tags=tags),
    **habitat_defaults(env),
)
trainer.run()
```

### Usage — VGGT variant (~70 lines, custom ObsAdapter)

```python
class VGGTObsAdapter(ObsAdapter):
    def __init__(self, extractor):
        super().__init__(buffer_dtype=np.float32, buffer_shape=(4116,),
                         normalize_on_sample=False, on_episode_reset=extractor.reset)
        self._extractor = extractor

    def transform(self, obs_dict):
        features = _flatten_vggt(self._extractor.extract(obs_dict["image"]))
        return features, {"features": features, "is_first": obs_dict.get("is_first", False)}

trainer = Trainer(agent=agent, env=env, agent_config=config,
                  trainer_config=TrainerConfig(...),
                  obs_adapter=VGGTObsAdapter(extractor),
                  episode_metrics_fn=habitat_defaults(env)["episode_metrics_fn"])
trainer.run()
```

### Usage — Crafter (~40 lines, no WandB, no callbacks)

```python
trainer = Trainer(agent=agent, env=env, agent_config=config,
                  trainer_config=TrainerConfig(output_dir=..., wandb_project=None))
trainer.run()
```

## What It Hides

| Concern | Before | After |
|---|---|---|
| Buffer selection | Caller picks `ReplayBuffer` vs `VGGTReplayBuffer` | `Trainer` creates unified buffer from `ObsAdapter` fields |
| `_convert_batch()` | 3 copies across scripts | Single `convert_batch()` called inside `Trainer.run()` |
| `_save_checkpoint()` | 2 copies, both miss `ema_state` | Single `save_checkpoint()` that includes `ema_state` |
| Training loop | ~100 lines × 3 scripts | Single implementation in `Trainer.run()` |
| CSV logging | Manual csv.writer in each script | Owned by `Trainer` |
| WandB init/log/finish | Inline in Habitat scripts, absent in Crafter | Conditional on `wandb_project` |
| Val loss | Only in `run_jax_habitat.py` | Conditional on `val_data` |
| Episode metrics | ~50 lines duplicated in Habitat + VGGT | `episode_metrics_fn` callback, pre-built by `habitat_defaults()` |

## Dependency Strategy

**In `trainer.py`:**
- Always imported: `jax`, `numpy`, `csv`, `pickle`, `os`, `time`
- Lazy (only if `wandb_project` is set): `wandb`, `EpisodeTracker`
- Lazy (only if `val_data` is set): `ValReplayDataset`

**Never imported by `trainer.py`:** `habitat`, `torch`, `DreamerConfig`. The env is constructed by the caller. The VGGT extractor lives in the caller's `ObsAdapter` subclass.

**Dependency category:** In-process (buffer, batch conversion, training loop) + Local-substitutable (checkpointing via pickle, logging via CSV/WandB).

## Testing Approach

**New boundary tests (replace smoke tests):**
- `test_replay_buffer.py`: Unified buffer with `obs_dtype="uint8"` and `obs_dtype="float32"` — verify storage, sampling, normalization. Replaces separate tests for both buffer classes.
- `test_convert_batch.py`: Verify one-hot encoding, field renaming. Pure function, no deps.
- `test_checkpoint_roundtrip.py`: `save_checkpoint → load_checkpoint → verify all state (including ema_state) matches`. Catches the current bug.
- `test_trainer_integration.py`: `Trainer` with a mock env (returns canned obs), mock agent (returns zeros), `wandb_project=None`. Verify: prefill runs, training steps happen, CSV file produced, checkpoint saved. No GPU/Habitat required.

**Can delete:**
- `VGGTReplayBuffer` class and its dedicated tests
- `_convert_batch()` copies in all three scripts
- `_save_checkpoint()` copies in two scripts

## Implementation Notes

**Order:**
1. Unify `ReplayBuffer` + `VGGTReplayBuffer` → single class with `BufferConfig`. Update existing callers.
2. Create `trainer.py` with `convert_batch()`, `save_checkpoint()`, `load_checkpoint()`, `ObsAdapter`, `TrainerConfig`, `habitat_defaults()`, `Trainer`.
3. Write tests for the new module.
4. Rewrite `run_jax_habitat.py` against the new `Trainer`. Verify identical training behavior (same seed → same metrics for N steps).
5. Rewrite `run_jax_habitat_vggt.py` with `VGGTObsAdapter`.
6. Rewrite `run_jax_crafter.py`.
7. Delete dead code (`VGGTReplayBuffer`, old `_convert_batch`/`_save_checkpoint` copies).

**Risks:**
- **Regression:** The existing scripts produced real experiment results. Must verify that refactored scripts produce identical training trajectories for the same seed before trusting them.
- **`DreamerConfig` still exists:** `HabitatObjectNavEnv` requires it. This RFC does not change that interface — it's a separate, smaller refactor to make the env accept keyword args instead.
- **`ObsAdapter.transform()` called once per step:** For VGGT, this means feature extraction happens inside `transform()`. If the caller needs the raw image AND the features (e.g., for visualization), they'd need to cache inside the adapter. Acceptable for now.

**What this RFC does NOT do:**
- Does not change `R2DreamerConfig` or `DreamerConfig` (config unification is deferred — the bridge is in `Trainer` construction)
- Does not change `R2DreamerAgent` or `HabitatObjectNavEnv` interfaces
- Does not add an env protocol/ABC (orthogonal refactor)

**Comment by LSailer** [comment]
## Implemented

Landed in 5 commits on `main`:

| Commit | Phase |
|---|---|
| `1a382da` | Unified `ReplayBuffer` + `VGGTReplayBuffer` with `BufferConfig` |
| `354a64a` | Created `modules/r2dreamer/trainer.py` (Trainer, convert_batch, checkpoint, ObsAdapter, habitat_defaults) |
| `5eb56e1` | Rewrote 3 training scripts against Trainer (758 → 291 lines, -62%) |
| `93a918d` | Review fixes: `tcfg.total_steps` drives loop, FPS logging restored, VGGT `--val_data` restored |
| `034ec6f` | Wiki: `docs/wiki/methods/training-orchestration.md` |

### Bug fixed
- `save_checkpoint()` now includes `agent.ema_state` — old code silently dropped return normalization state on resume.

### Tests added
- 12 tests for unified `ReplayBuffer` (uint8, float32, normalization, wrapping, is_first)
- 8 tests for `convert_batch` (5) and checkpoint round-trip (3, including ema_state verification)

### Pre-existing issues surfaced by review (not fixed here)
- `ReplayBuffer.sample()` can cross the write head after buffer wraps — silently blends non-adjacent transitions
- `is_first[:, 0] = True` unconditionally resets RSSM state even for mid-episode sequences
- `ValReplayDataset.sample()` hardcodes `/255.0` — would break VGGT float32 validation data
